File: LandLord.Engine/Bomb.fs
namespace Itminus.LandLord.Engine

open Card

module Bomb = 
    
    let (|Bomb|_|) (cards: PlayingCard list)= 
        match cards with
        | [NormalCard(v1, _); NormalCard(v2, _); NormalCard(v3, _); NormalCard(v4, _) ]
            when v1 = v2 && v2 = v3 && v3 = v4 -> Some(cards)
        | [Joker(j1); Joker(j2)] 
            when j1 <> j2 -> Some(cards)
        | _ -> None
    

File: LandLord.Core/Patterns/Bomb.cs
using LandLord.Shared;
using System.Collections.Generic;

namespace LandLord.Core.Patterns
{
    public static partial class Patterns
    {
        public static (bool, IList<PlayingCard>) Bomb(IList<PlayingCard> cards)
        {
            if(cards.Count <2) return (false,null);
            var first = cards[0];
            var second= cards[1];
            if(cards.Count == 2)
            {
                switch((first, second)){
                    case (JokerCard j1, JokerCard j2) 
                        when j1 != j2:
                        return (true, cards);
                    default:
                        return (false,null);
                }
            }
            else if(cards.Count == 4)
            {
                var third = cards[2];
                var forth = cards[3];
                switch((first, second, third, forth))
                {
                    case (NormalCard { CardValue: var cv1 }, NormalCard { CardValue: var cv2 }, NormalCard { CardValue: var cv3 }, NormalCard { CardValue: var cv4 })
                        when cv1 == cv2 && cv2 == cv3 && cv3 == cv4:
                        return (true,cards);
                    default:    
                        return (false,null);
                }
            }
            return (false,null);
        }
    }
}